dotnet-desktop-guide/winforms/migration/clipboard-dataobject-net10.md

```diff
@@ -0,0 +1,394 @@
+---
+title: "Clipboard and DataObject changes in .NET 10"
+description: "Learn about the major clipboard and drag-and-drop changes in .NET 10, including new type-safe APIs, JSON serialization, and migration from BinaryFormatter."
+author: adegeo
+ms.author: adegeo
+ms.service: dotnet-desktop
+ms.topic: concept-article
+ms.date: 09/26/2025
+dev_langs:
+  - "csharp"
+  - "vb"
+ms.custom:
+  - copilot-scenario-highlight
+ai-usage: ai-assisted
+
+#customer intent: As a Windows Forms developer, I want to understand the clipboard and DataObject changes in .NET 10 so that I can upgrade my applications and use the new type-safe APIs.
+
+---
+
+# Windows Forms clipboard and DataObject changes in .NET 10
+
+This article shows you how to upgrade your Windows Forms clipboard and drag-and-drop operations to the new type-safe APIs in .NET 10. You'll learn how to use the new <xref:System.Windows.Forms.Clipboard.TryGetData*?displayProperty=nameWithType> and <xref:System.Windows.Forms.Clipboard.SetDataAsJson``1(System.String,``0)?displayProperty=nameWithType> methods, understand which built-in types work without changes, and discover strategies for handling custom types and legacy data after the removal of `BinaryFormatter`.
+
+`BinaryFormatter` was removed from the runtime in .NET 9 because of security vulnerabilities. This change broke clipboard and drag-and-drop operations with custom objects. .NET 10 introduces new APIs that use JSON serialization and type-safe methods to restore this functionality, improve security, and provide better error handling and cross-process compatibility.
+
+One significant change is that <xref:System.Windows.Clipboard.SetData(System.String,System.Object)?displayProperty=nameWithType> no longer works with custom types. It silently fails without storing data on the clipboard. <xref:System.Windows.Forms.Clipboard.GetData(System.String)?displayProperty=nameWithType> is obsolete in .NET&nbsp;10 and shouldn't be used, even for built-in types. Use the new <xref:System.Windows.Forms.Clipboard.TryGetData*?displayProperty=nameWithType> and <xref:System.Windows.Forms.Clipboard.SetDataAsJson``1(System.String,``0)?displayProperty=nameWithType> methods for type-safe operations and JSON serialization of custom objects.
```

Comment:
No longer works "by default".

To see the failure, you have to catch first chance exceptions under the debugger to see it fail on OLE trying to retrieve a copy of the data (which happens when you set if `copy == true`, the default, otherwise on get).